Feature/ROC-78

yakimk opened the pull request:
Simple endpoint, I tried to do it in a similar fashion as before.  Also added a basic test that is based on what  Bartek has written for other endpoints. Tested it by running a couple requests to db, it seems to work correctly.

pawlowiczf commented:
Może warto przenieść folder `match`, gdzieś indziej (z folderu `user`) ? Po za tym, wydaje się ok, ale poczekajmy najpierw na review kogoś innego.

LisieckaK00 commented on backend/src/main/java/com/roc/app/user/match/Match.java:
```diff
@@ -0,0 +1,111 @@
+    @Column(name = "winner_id")
+    private Integer winnerId;
+
+    public Integer getMatchId() {
```
Please remove all getters and setters and replace them with `@Data` annotation on this class.

LisieckaK00 commented on backend/src/main/java/com/roc/app/user/match/UserMatchController.java:
```diff
@@ -0,0 +1,22 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("/users/{userId}/matches")
```
We're duplicating the paths here. This class path should be `/api/matches` and here should be methods for adding, updating and deleting matches.

LisieckaK00 commented on backend/src/main/java/com/roc/app/user/match/UserMatchController.java:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    @GetMapping
+    public List<UserMatchDto> getUserMatches(@PathVariable Integer userId) {
```
This method should indeed have path like `/api/users/{userId}/matches` but it should be located in the `UsersController`.

LisieckaK00 commented on backend/src/main/java/com/roc/app/user/match/UserMatchController.java:
```diff
@@ -0,0 +1,22 @@
+
+@RestController
+@RequestMapping("/users/{userId}/matches")
+public class UserMatchController {
```
We have `Match` in model and so here we should have just `MatchController`. Same applies to service and repository.
```suggestion
public class MatchController {
```

LisieckaK00 commented on backend/src/main/java/com/roc/app/user/match/dto/UserMatchDto.java:
```diff
@@ -0,0 +1,30 @@
+
+@Getter
+@Setter
+public class UserMatchDto {
```
In this project we use `record` instead of a `class` as it comes with all getters, setters and constructor.

LisieckaK00 commented on backend/src/main/java/com/roc/app/user/match/dto/UserMatchDto.java:
```diff
@@ -0,0 +1,30 @@
+@Getter
+@Setter
+public class UserMatchDto {
+    private Integer matchId;
```
Add `@NotNull` and `@NotBlank` to all the fields. You may see for example `CompetitionCreateRequestDto` for reference.

LisieckaK00 commented on backend/src/main/java/com/roc/app/user/match/MatchStatus.java:
I think it'll be easier for us to remove this table in the database and make this class an `enum`. It requires to remove from the sql file all code related to `match_statuses`. Currently, there are inserts, so you may see what statuses we thought of.

LisieckaK00 commented on backend/src/main/java/com/roc/app/user/match/CompetitionParticipant.java:
```diff
@@ -0,0 +1,38 @@
+    private Integer participantId;
+
+    @Column(name = "status_id")
+    private Integer statusId;
```
Same situation here, let's make this an enum, and get rid of `participant_statuses`.

LisieckaK00 reviewed (changes requested):
I'be just added some comments about the database schema. It's probably easier to make this changes than deal with this additional tables. Sorry it wasn't designed like this from the beginning, didn't think it'll be a problem back then.

yakimk commented:
So I added the changes and merged main. I am not sure why, but two tests in `AppApplicationTests`, which are actually empty, are failing now.
```java
@SpringBootTest
class AppApplicationTests {

	@Test
	void contextLoads() {
	}

	@Test
	void trueTest() {
        assertTrue(true);
	}
}
```

1. I haven't implemented `MatchController` and `MatchService`. I think we would need to do them on a separate branch. 
2. I am waiting for a db update to properly modify the query in `UserMatchRepository`.

LisieckaK00 commented on backend/src/main/java/com/roc/app/match/UserMatchService.java:
```diff
@@ -0,0 +1,20 @@
+import java.util.List;
+
+@Service
+public class UserMatchService {
```
```suggestion
public class MatchService {
```

LisieckaK00 commented on backend/src/main/java/com/roc/app/match/UserMatchService.java:
```diff
@@ -0,0 +1,20 @@
+        this.matchRepository = matchRepository;
+    }
+
+    public List<UserMatchDto> getUserMatches(Integer userId) {
```
We need to differentiate participants and referees, as we'll be looking for these matches differently. If I understand correctly, current logic is about a participant. So please replace everywhere usage of `user` with `participant`, like `getPartcipantMatches`, `findParticipantMatches` and so on.

LisieckaK00 commented on backend/src/main/java/com/roc/app/match/dto/UserMatchDto.java:
```diff
@@ -0,0 +1,31 @@
+
+import java.time.LocalDateTime;
+
+public record UserMatchDto(
```
```suggestion
public record ParticipantMatchResponseDto(
```
Verify with the database. Some field may be null, as we'll be showing the participant matches, that didn't occur yet. Then, we won't know the score or winner. In the beginning, we won't even know the opponent. Please adjust accordingly.

LisieckaK00 commented on backend/src/main/java/com/roc/app/match/MatchController.java:
```diff
@@ -0,0 +1,40 @@
+@RequestMapping("/api/matches")
+public class MatchController {
+
+// Will have to do match service on a separate branch, so commented out for now.
```
We don't have a separate task for this, I believe we should add this now.

LisieckaK00 commented on backend/src/main/java/com/roc/app/user/general/UserController.java:
```diff
@@ -25,4 +31,9 @@ public ResponseEntity<UserResponseDto> createUser(@Valid @RequestBody UserCreate
         return ResponseEntity.ok(user);
     }
+
+    @GetMapping("/{userId}/matches")
```
Sorry, I didn't notice one thing when I was using the review before. This is for a participant and not the user in general. Referee will have different query. Please, move this to participant.

LisieckaK00 commented on backend/src/main/java/com/roc/app/match/UserMatchRepository.java:
```diff
@@ -0,0 +1,34 @@
+
+import java.util.List;
+
+public interface UserMatchRepository extends Repository<Match, Integer> {
```
```suggestion
public interface MatchRepository extends Repository<Match, Integer> {
```

LisieckaK00 reviewed (changes requested):
As I wrote in some of my comments, current logic doesn't support user in general, but only the participant. Please also add logic that will allow seeing matches for a referee.

LisieckaK00 commented on backend/src/main/java/com/roc/app/match/CompetitionParticipant.java:
```diff
@@ -0,0 +1,40 @@
+public class CompetitionParticipant {
+
+    @Id
+    @Column(name = "competition_id")
```
Please move column names to a class, as in `Match`.

LisieckaK00 commented on backend/src/main/java/com/roc/app/match/CompetitionParticipant.java:
As this entity is connected to participant and match, I recommend creating a new package. Maybe we could create a package `competitionassignment`, and place there this class, `ParticipantStatus` and `CompetitonReferee` that we'll create in the future. I don't think that `match` package is a good place for this class.

yakimk commented:
- added suggested changes
- added update/create/remove match functionality
- added new tests
- all tests pass
- wasn't able to test it with a db. Couldn't start a docker container because of some problem on forntend

LisieckaK00 commented on backend/src/main/java/com/roc/app/match/dto/ParticipantMatchResponseDto.java:
```diff
@@ -0,0 +1,28 @@
+        @NotBlank(message = "Status is required")
+        String status,
+
+        String opponentName,
```
Based on our figma design, we'll also need referee name here.
